Avoid wxEVT_TEXT_ENTER when completion popup is shown in wxGTK

AliKet opened the pull request:
This is actually what this  pull request tries to do https://github.com/wxWidgets/wxWidgets/pull/728
but takes another approach to solve the same problem.

For consistency with wxMSW, and also because it just makes more sense
from UI point of view, avoid generating wxEVT_TEXT_ENTER events when an
auto-completion popup is shown and just accept the current selection in
it when Enter is pressed instead.

This notably allows to use "Enter" naturally with wxSearchCtrl using
auto-completion, instead of generating wxEVT_SEARCH events when the user
just wants to select a popup entry.

In fact, wxTextCtrl has the same issue if you bind wxEVT_TEXT_ENTER event
to a custom handler which prevents the popup from dismissing itself if you
don't Skip() the event. Or it results in the custom handler being run before the selection
from the popup takes place if you do.

Commit:
Avoid wxEVT_TEXT_ENTER when completion popup is shown in wxGTK

vadz commented:
@paulcor I think this is preferable to #728, which solves the same problem, what do you think?

paulcor commented:
Yes, looks preferable. Seems like it could be simplified, see below (no comments for brevity):
```
diff --git a/src/gtk/textentry.cpp b/src/gtk/textentry.cpp
index 84144d2289..1bd278198c 100644
--- a/src/gtk/textentry.cpp
+++ b/src/gtk/textentry.cpp
@@ -195,6 +195,7 @@ wx_gtk_paste_clipboard_callback( GtkWidget *widget, wxWindow *win )
         widget, win, wxEVT_TEXT_PASTE, "paste-clipboard" );
 }
 
+static void grab_notify(GtkWidget* widget, gboolean was_grabbed, wxWindow* win);
 } // extern "C"
 
 // Base class for wxTextAutoCompleteFixed and wxTextAutoCompleteDynamic below.
@@ -213,6 +214,7 @@ public:
     void DisableCompletion()
     {
         gtk_entry_set_completion (GetGtkEntry(), NULL);
+        g_signal_handlers_disconnect_by_data(GetGtkEntry(), this);
     }
 
     virtual ~wxTextAutoCompleteData()
@@ -238,6 +240,12 @@ protected:
 
         gtk_entry_completion_set_text_column (completion, 0);
         gtk_entry_set_completion (GetGtkEntry(), completion);
+        wxWindow* editableWin = GetEditableWindow(m_entry);
+        if (editableWin->HasFlag(wxTE_PROCESS_ENTER))
+        {
+            g_signal_connect(GetGtkEntry(), "grab-notify",
+                G_CALLBACK(grab_notify), editableWin);
+        }
     }
 
     // Provide access to wxTextEntry::GetEditableWindow() to the derived
@@ -410,6 +418,17 @@ private:
     wxDECLARE_NO_COPY_CLASS(wxTextAutoCompleteDynamic);
 };
 
+extern "C" {
+static void grab_notify(GtkWidget* widget, gboolean was_grabbed, wxWindow* win)
+{
+    long flag = win->GetWindowStyleFlag();
+    if (gtk_widget_has_focus(widget) && !was_grabbed)
+        flag &= ~wxTE_PROCESS_ENTER;
+    else
+        flag |= wxTE_PROCESS_ENTER;
+    win->SetWindowStyleFlag(flag);
+}
+}
 
 // ============================================================================
 // wxTextEntry implementation
```

AliKet commented:
I think this line :
        g_signal_handlers_disconnect_by_data(GetGtkEntry(), this);

should be :
        wxWindow* editableWin = GetEditableWindow(m_entry);
        if (editableWin->HasFlag(wxTE_PROCESS_ENTER))
        {
            g_signal_handlers_disconnect_by_data(GetGtkEntry(), editableWin);
        }
?

paulcor commented:
Oops, yes. The `if` is not necessary though.

vadz committed:
Minor cleanup after the previous commit

Make m_origWinFlags const; keep just a single call to
SetWindowStyleFlag() and some space harmonization.

No real changes.

vadz commented:
I actually don't think simplification is totally correct, as `wxTE_PROCESS_ENTER` can be turned on later, after setting up the completion, and even if it's a slightly esoteric situation, I don't see any real reason why this shouldn't be allowed -- and it's not like we simplify the code that much anyhow. So I'll merge the original patch (thanks @AliKet!) with just some tiny cleanup.

FWIW I still think that ideally we'd avoid generating `CHAR` events when Enter is pressed in the popup, but this is definitely better than nothing and should be good enough in practice.

AliKet commented:
Looks perfect now! Thanks.

paulcor commented:
> I actually don't think simplification is totally correct, as wxTE_PROCESS_ENTER can be turned on later, after setting up the completion

In which case the version you committed won't work either, as _all_ window flags changed after setting up the completion will be overwritten at the first keypress. Testing with the patch below will verify this.
```
diff --git a/src/generic/srchctlg.cpp b/src/generic/srchctlg.cpp
index 074bee8722..083510ca35 100644
--- a/src/generic/srchctlg.cpp
+++ b/src/generic/srchctlg.cpp
@@ -823,7 +823,11 @@ bool wxSearchCtrl::SetBackgroundColour(const wxColour& colour)
 // Autocomplete
 bool wxSearchCtrl::DoAutoCompleteStrings(const wxArrayString &choices)
 {
-    return m_text->AutoComplete( choices );
+    long flag = m_text->GetWindowStyle();
+    m_text->SetWindowStyle(flag & ~wxTE_PROCESS_ENTER);
+    bool b = m_text->AutoComplete( choices );
+    m_text->SetWindowStyle(flag);
+    return b;
 }
 
 bool wxSearchCtrl::DoAutoCompleteFileNames(int flags)
```

AliKet commented:
Hi
Sorry, but your test case is not correct ! as m_text is an implementation detail and by design should have
wxTE_PROCESS_ENTER flag always turned on. and tampering with this invariant has nothing to do with auto-completion.

paulcor commented:
> Sorry, but your test case is not correct ! as m_text is an implementation detail

It demonstrates that the exact scenario @vadz described ("`wxTE_PROCESS_ENTER` can be turned on later, after setting up the completion") doesn't work in the code that was committed. That it uses the internals of `wxSearchCtrl` to do it is irrelevant, and not the point.

vadz commented:
Yes, @paulcor is absolutely correct, I was thinking about the first version of the patch when discussing the second one and the result, unsurprisingly, doesn't make any sense. I'll fix it soon.
